### ISSUE TITLE
[MD] "failure in decrypting message"

### ISSUE THREAD

**Issue by burstfreeze** [adiwajshing/Baileys#744]
Similar to #735 #704 #697

I also saw a "failure in decrypting message" now. It does not mess up the whole session, just particular conversations, others are fine. Since some people have reported this on session restarts and me having this issue after launching again I suppose it has something to do with WhatsApp expecting keys to be stored - keys that never arrived at the client side. 

For example if someone sends a message during session restart.

Just my 2 thoughts on this.

**Comment by burstfreeze** [comment]
Don't mind me, I'm just a bump

**Comment by burstfreeze** [comment]
@adiwajshing should the group chats that one is in touch with ALL have entries in the auth json? If so, then that is the root of the issue - the group that suddenly does not receive messages anymore is not mentioned anywhere in the auth.json

**Comment by burstfreeze** [comment]
One idea:

Since I am of the opinion that this tends to happen during session restarts there is something that came to my mind.

https://github.com/adiwajshing/Baileys/blob/multi-device/src/Socket/messages-recv.ts#L386

Are all these receipts for message reception or is at **least one of them** a "I received the new keys" receipt? If so, and if it is supposed to save the updated creds afterwards there is something could arise

The loop sends a receipt for each message received HOWEVER the cred update only happens once all of these were sent (the sending blocks the execution for a tiny amount of time). If the process is stopped in that loop, we tell the server that we have received keys that we never had the chance to store.

I would make a PR but if you can confirm whether that is what happens or at least **could** happen that would be golden. The most idiotic way I can think of is to put save events on each message or work through them in smaller batches to avoid expensive disk load and only then send the receipts.

Again I may be completely wrong, just a wild guess that at first makes sense to me.

**Comment by rejidan** [comment]
maybe this can help, I can decrypt the messages received after restarting by updating the state in 'messages.update' :
 
```
   // save the authentication state to a file
    const saveState = (state?: any) => {
        console.log('saving pre-keys')
        state = sock?.authState
        writeFileSync(
            './auth_info_multi.json',
            // BufferJSON replacer utility saves buffers nicely
            JSON.stringify(state, BufferJSON.replacer, 2)
        )
    }

        sock.ev.on('messages.update', m => {
            let msg = m[0];
            if (msg.key.remoteJid !== 'status@broadcast') {
                saveState()
            }
        })
```

**Comment by burstfreeze** [comment]
> maybe this can help, I can decrypt the messages received after restarting by updating the state in 'messages.update' :

Thank you for sharing @rejidan - seems like a lot of overhead considering fs access but I will try that. 

So the flow is as following - you notice the error, cancel the process and `npm run` again? Or do you have an auto-reconnect set in place?

**Comment by burstfreeze** [comment]
Unfortunately it did not fix it for me. Putting the `save()`s in there does not help.

**Comment by burstfreeze** [comment]
On the current commit I get `Received message with old counter: 3, 1` shortly before.

**Comment by burstfreeze** [comment]
What's interesting is that after each restart of the session that cannot receive a message, we see `Error: No sender key for: xxx-yyy@g.us::xxx:1::0`

My new guess: The key is missing, the message cannot be decrypted, we therefore do not send a delivery receipt and the server will not send any more messages to us unless we decrypt and ack the message that's causing it.

**Comment by burstfreeze** [comment]
I have just noticed something.

Using `.sendRawMessage()` ws was null. May be my fault, maybe not - could there be an instance where Baileys tries to ACK but can't because for some reason ws broke down? 

Now to the good news, after restarting I saw some `fetching session events` shortly before I tried to send a message to the stuck chat.

Then I received a senderKeyDistributionMessage, `sent receipt, ack, "reading messages", delivery receipt, message receipt, message ack, "reading messages", delivery receipt `, same for another chat which I don't know whether it was stuck and everything went back to normal.

In other words, I now had a session recover from this, dunno if today's commits did that :)

**Comment by diveshahuja** [comment]
I am having the same issue, "failure in decrypting message"

**Comment by textarea** [comment]
same problem on few accounts
